File: ADB Date Changer/ViewController.m
//
//  ViewController.m
//  ADB Date Changer
//
//  Created by Vitaliy on 5/6/16.
//  Copyright © 2016 vitovalov. All rights reserved.
//

#import "ViewController.h"

@implementation ViewController


- (IBAction)onCalend:(NSDatePicker *)sender {
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    [dateFormatter setDateFormat:@"yyyyMMdd.HHmmss"];
    
    NSDate *vas = [_calendar dateValue];
    
    NSString *formattedDate = [dateFormatter stringFromDate:vas];
    
    NSString* filePath = [[NSBundle mainBundle] pathForResource:@"adb" ofType:@""];
    //NSURL *resourcesUrl = [NSURL fileURLWithPath:filePath];

    NSString *greeting = [[NSString alloc] initWithFormat:@"\"%@\"",filePath];
    NSString *a = [[NSString alloc] initWithFormat:@" -e shell \'su 0 date -s %@",formattedDate];
    NSString *b = [a stringByAppendingString:@"\'"];
                   
    NSString * dos = [greeting stringByAppendingString:b];
    
    /*
    int pid = [[NSProcessInfo processInfo] processIdentifier];
    NSPipe *pipe = [NSPipe pipe];
    NSFileHandle *file = pipe.fileHandleForReading;
    
    NSTask *task = [[NSTask alloc] init];
    task.launchPath = @"/bin/sh";
    task.arguments = @[@"-c",dos];
    task.standardOutput = pipe;
    
    [task launch];
    
    NSData *data = [file readDataToEndOfFile];
    [file closeFile];
    
    NSString *output = [[NSString alloc] initWithData: data encoding: NSUTF8StringEncoding];
    NSLog (@"adb returned:\n%@", output);
*/
    
    NSString *output = [self runCommand:dos];
}

- (NSString *)runCommand:(NSString *)commandToRun
{
    NSTask *task = [[NSTask alloc] init];
    [task setLaunchPath:@"/bin/sh"];
    
    NSArray *arguments = [NSArray arrayWithObjects:
                          @"-c" ,
                          [NSString stringWithFormat:@"%@", commandToRun],
                          nil];
    NSLog(@"run command:%@", commandToRun);
    [task setArguments:arguments];
    
    NSPipe *pipe = [NSPipe pipe];
    [task setStandardOutput:pipe];
    
    NSFileHandle *file = [pipe fileHandleForReading];
    
    [task launch];
    
    NSData *data = [file readDataToEndOfFile];
    
    NSString *output = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    return output;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
}


- (void)setRepresentedObject:(id)representedObject {
    [super setRepresentedObject:representedObject];
}

@end

File: ADB Date Changer/ViewController.swift
//
//  ViewController.swift
//  ADB Date Changer
//
//  Created by Vitaliy Konovalov on 4/6/16.
//  Copyright © 2016 Vitovalov. All rights reserved.
//

import Cocoa

class ViewController: NSViewController {

    @IBOutlet weak var calendar: NSDatePickerCell!
    
    @IBAction func calendarPicked(sender: AnyObject) {
        
        print(calendar.dateValue.description)
        
        let dateFormatter = NSDateFormatter()
        dateFormatter.dateFormat = "yyyyMMdd.HHmmss"
        let formattedDate = dateFormatter.stringFromDate(calendar.dateValue)
        
        let resourcesUrl = NSURL(fileURLWithPath: NSBundle.mainBundle().pathForResource("adb", ofType: "")!).URLByStandardizingPath
        let resourcesPath = resourcesUrl?.path
        
        let jaf = "'"+resourcesPath! + "'"
        let dos = jaf + " -e shell \'su 0 date -s "+formattedDate + "\'"
        
        print(dos)
        
        let output = dos.runAsCommand()
        
        print(output)
    }
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
    }

    override var representedObject: AnyObject? {
        didSet {
        // Update the view, if already loaded.
        }
    }


}


extension String {
    func runAsCommand() -> String {
        let pipe = NSPipe()
        let task = NSTask()
        task.launchPath = "/bin/sh"
        task.arguments = ["-c", String(format:"%@", self)]
        task.standardOutput = pipe
        let file = pipe.fileHandleForReading
        task.launch()
        if let result = NSString(data: file.readDataToEndOfFile(), encoding: NSUTF8StringEncoding) {
            return result as String
        }
        else {return "--- Unable to initialize string from file data ---"}
    }
}
